examples/tutorials/matmul/int4_weight_decompression.cpp

```diff
@@ -0,0 +1,265 @@
+///
+/// @page int4_weights_decompression_matmul_cpp MatMul Tutorial: weights
+/// decompression
+/// @copydetails int4_weights_decompression_matmul_cpp
+///
+/// Assumptions:
+/// 1. The shape of the weights (matrix \f$B(K, N)\f$) is known in advance, the
+///    data type is `int4` and shifted from 0 (i.e. the zero point is not 0).
+/// 2. The source matrix \f$A\f$ and destination matrix \f$C\f$ have floating
+///    point data type.
+/// 3. Scaling (re-quantization) factor specified at run-time only.
+///
+/// Since the shape of weights is known in advance, the MatMul weights can be
+/// created with format tag #dnnl::memory::format_tag::any to enable the library
+/// to choose the most appropriate layout for best performance.
+///
+/// @warning
+/// The format tag #dnnl::memory::format_tag::any doesn't work for memory
+/// descriptors that have one or more unknown dimensions and/or strides.
+///
+/// @include int4_weight_decompression.cpp
+#include <cassert>
+#include <cctype>
+#include <cmath>
+#include <cstdio>
+#include <iostream>
+#include <random>
+#include <stdexcept>
+#include <vector>
+
+#include "oneapi/dnnl/dnnl.hpp"
+
+#include "example_utils.hpp"
+
+using namespace dnnl;
+
+namespace {
+
+void init_vector(std::vector<float> &v) {
+    std::mt19937 gen;
+    std::uniform_real_distribution<float> u(0, 1);
+    for (auto &e : v)
+        e = u(gen);
+}
+// Comparing two vectors by calculating their L2 norms and the L2 norm of their
+// difference Checking if the difference is within a calculated threshold The
+// function returns 0 if the vectors are considered similar, otherwise it
+// returns 1.
+int compare_vectors(const std::vector<float> &v1, const std::vector<float> &v2,
+        int64_t K, const char *message) {
+    double v1_l2 = 0, diff_l2 = 0;
+    for (size_t n = 0; n < v1.size(); ++n) {
+        float diff = v1[n] - v2[n];
+        v1_l2 += v1[n] * v1[n];
+        diff_l2 += diff * diff;
+    }
+
+    v1_l2 = std::sqrt(v1_l2);
+    diff_l2 = std::sqrt(diff_l2);
+
+    // Finding the reasonable (tight and accurate) threshold is quite difficult
+    // problem.
+    // The implementation testing might also use special data filling to
+    // alleviate issues related to the finite precision arithmetic.
+    // However, in simple cases the machine epsilon multiplied by log(K) should
+    // work reasonably well.
+    const double threshold = std::numeric_limits<float>::epsilon()
+            * std::log(std::max(2., (double)K));
+    bool ok = diff_l2 <= threshold * v1_l2;
+
+    printf("%s\n\tL2 Norms"
+           "\n\t\tReference matrix:%g\n\t\tError:%g\n\t\tRelative_error:%g\n"
+           "\tAccuracy check: %s\n",
+            message, v1_l2, diff_l2, diff_l2 / v1_l2, ok ? "OK" : "FAILED");
+
+    return ok ? 0 : 1;
+}
+
+} // namespace
+
+// Floating point MatMul
+// Inputs:
+// - Shape: M, N, K
+// - Matrices A and B
+// Outputs:
+// - Matrix C
+void ref_compute_matmul_f32(int64_t M, int64_t N, int64_t K, int64_t G,
```

Comment:
I would suggest dropping fp32 reference comparison from this example as it does not add value when explaining int4 quantization.